src/boot/efi/stub.c

```diff
@@ -15,8 +16,68 @@
 /* magic string to find in the binary image */
 static const char __attribute__((used)) magic[] = "#### LoaderInfo: systemd-stub " GIT_VERSION " ####";
 
+static EFI_STATUS combine_initrd(
+                EFI_PHYSICAL_ADDRESS initrd_base, UINTN initrd_size,
+                const VOID *credential_initrd, UINTN credential_initrd_size,
+                const VOID *sysext_initrd, UINTN sysext_initrd_size,
+                EFI_PHYSICAL_ADDRESS *ret_initrd_base, UINTN *ret_initrd_size) {
+
+        EFI_PHYSICAL_ADDRESS base = (EFI_PHYSICAL_ADDRESS) -1;
+        EFI_STATUS err;
+        UINT8 *p;
+        UINTN n;
+
+        /* Combines three initrds into one, by simple concatenation in memory */
+
+        n = initrd_size;
+        if (credential_initrd) {
+                if (n > UINTN_MAX - credential_initrd_size)
+                        return EFI_OUT_OF_RESOURCES;
+
+                n += credential_initrd_size;
+        }
+        if (sysext_initrd) {
+                if (n > UINTN_MAX - sysext_initrd_size)
+                        return EFI_OUT_OF_RESOURCES;
+
+                n += sysext_initrd_size;
+        }
+
+        err = uefi_call_wrapper(
+                        BS->AllocatePages, 4,
+                        AllocateMaxAddress,
+                        EfiLoaderData,
+                        EFI_SIZE_TO_PAGES(n),
+                        &base);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to allocate space for combined initrd: %r", err);
+
+        p = (UINT8*) (UINTN) base;
+
+        if (credential_initrd) {
+                CopyMem(p, credential_initrd, credential_initrd_size);
+                p += credential_initrd_size;
+        }
+
+        if (sysext_initrd) {
+                CopyMem(p, sysext_initrd, sysext_initrd_size);
+                p += sysext_initrd_size;
+        }
+
+        if (initrd_base != 0) {
+                CopyMem(p, (VOID*) (UINTN) initrd_base, initrd_size);
```

Comment:
initrd_base last? Will this work with early microcode loading?

Reply:
I just checked this and the kernel does *not* do early (or any) microcode loading if the microcode cpio is not the first. And that one would normally be at the start of initrd (which itself is already a cat of 2 archives, one uncompressed and one compressed).